### ISSUE TITLE
New rule: constant classes

### ISSUE THREAD

**Issue by FreHu** [abaplint/abaplint#2125]
if you have a domain with fixed values, specify a class which serves as constants of those values

config
```json
"constant_classes": {
  "mapping": [ {"blah": "cl_blah_constants"}, ... ]
}
```

so for domain `blah` with values `foo`, `bar`, there must exist a `cl_blah_constants` with the constants in the public section

```abap
constants foo type blah value `foo`.
constants bar type blah value `bar`.
```

- the check would detect missing values, extra values and mismatches
- a quickfix would be able to implement the constants (or the whole class, so --fix would generate code for you)

**Comment by larshp** [comment]
👍 why not

**Comment by mbtools** [comment]
I like the idea. In general, a rule to validate the mapping between DDIC and code would be good to have (DOMA > constant, DTEL > data type, TABL > structure). So many reconsider the name of the rule. 

In the case of fixed domain values, I think it's more common to use a constant structure than individual constants and it should work with interfaces, too. 

```abap
  CONSTANTS:
    BEGIN OF c_method,
      add    TYPE zmethod VALUE 'A',
      rm     TYPE zmethod VALUE 'R',
      ignore TYPE zmethod VALUE 'I',
      skip   TYPE zmethod VALUE '?',
    END OF c_method.
```

**Comment by FreHu** [comment]
Constant structures are valid, but usage on interfaces is not recommended in the guidelines
https://github.com/SAP/styleguides/blob/main/clean-abap/CleanABAP.md#prefer-enumeration-classes-to-constants-interfaces

My suggestion is for the rule to also work if you specify the constant structure name
```
"constant_classes": {
  "mapping": [ {"blah": "cl_blah_constants=>c_method"}, ... ]
}
```
or separate properties for class name and structure name. The rule could be renamed to something like `domain_value_constants`

**Comment by mbtools** [comment]
Yes, I read https://github.com/SAP/styleguides/blob/main/clean-abap/sub-sections/Enumerations.md again. Your original suggestion is for checking the "constant pattern". It's the Clean ABAP recommendation which is fine and we should *not* add the constant structure ("collection pattern"). To be "clean" the rule should check that there are no other constants in this class ;-). Maybe `domain_constant_pattern`?

PS: I had been thinking more about checking the completeness of "interface/collection patterns". Maybe that's another rule that identifies these patterns and provides a quick fix to turn them into classes with "constant pattern".

**Comment by larshp** [comment]
lots more stuff can be done in this area

small steps, implementing the `constant_class` rule will get some of the scaffolding in place, which can be used to make more complex rules

**Comment by FreHu** [comment]
I don't mind the collection pattern but what we use currently for domain values is the separate classes so i went with that first. In this pattern we sometimes have:
- static methods which return something about the domain, like validation of value or get_shorttext
- extra constant which contains the domain name

Multiple types in a single class is something i also often do, just not for enumerated values. Pretty much all of my methods nowadays have a t_input and t_output type so I can avoid needing multiple parameters.

I can see this rule 
- being renamed to `constant_class_pattern` and having a separate rule for `collection_pattern` 
or 
- adding another property to specify the collection name where the constants are supposed to be

from the technical side 
- i added parsing of domain values and constants, but not sure if grouped constants are available in the file info yet
- quickfix to add a constant collection should be easier than to implement than to create a whole new class
